base_layer/key_manager/src/cipher_seed.rs

```diff
@@ -155,223 +155,205 @@ impl CipherSeed {
-        let mut enciphered_seed = body.split_off(1);
-        let received_version = body[0];
-
-        // apply cipher stream
-        Self::apply_stream_cipher(&mut enciphered_seed, &passphrase, salt.as_slice())?;
-
-        let decrypted_mac = enciphered_seed.split_off(enciphered_seed.len() - CIPHER_SEED_MAC_BYTES);
+        // Derive encryption and MAC keys from passphrase and main salt
+        let passphrase = Zeroizing::new(passphrase.unwrap_or_else(|| DEFAULT_CIPHER_SEED_PASSPHRASE.to_string()));
+        let salt = encrypted_seed
+            .split_off(1 + CIPHER_SEED_BIRTHDAY_BYTES + CIPHER_SEED_ENTROPY_BYTES + CIPHER_SEED_MAC_BYTES);
+        let (encryption_key, mac_key) = Self::derive_keys(&passphrase, &salt)?;
 
-        let decrypted_entropy_vec: ArrayVec<_, CIPHER_SEED_ENTROPY_BYTES> =
-            enciphered_seed.split_off(2).into_iter().collect();
-        let decrypted_entropy = decrypted_entropy_vec
-            .into_inner()
-            .map_err(|_| KeyManagerError::InvalidData)?;
+        // Decrypt the secret data: birthday, entropy, MAC
+        let mut secret_data = Zeroizing::new(encrypted_seed.split_off(1));
+        Self::apply_stream_cipher(&mut secret_data, encryption_key.as_ref(), &salt)?;
 
-        let mut birthday_bytes: [u8; CIPHER_SEED_BIRTHDAY_BYTES] = [0u8; CIPHER_SEED_BIRTHDAY_BYTES];
-        birthday_bytes.copy_from_slice(&enciphered_seed);
-        let decrypted_birthday = u16::from_le_bytes(birthday_bytes);
+        // Parse secret data
+        let mac = secret_data.split_off(CIPHER_SEED_BIRTHDAY_BYTES + CIPHER_SEED_ENTROPY_BYTES);
+        let entropy = Zeroizing::new(secret_data.split_off(CIPHER_SEED_BIRTHDAY_BYTES));
+        let mut birthday_bytes = [0u8; CIPHER_SEED_BIRTHDAY_BYTES];
+        birthday_bytes.copy_from_slice(&secret_data);
+        let birthday = u16::from_le_bytes(birthday_bytes);
 
-        // generate the MAC
-        let mac = Self::generate_mac(
-            &decrypted_birthday.to_le_bytes(),
-            &decrypted_entropy,
-            &[CIPHER_SEED_VERSION],
-            salt.as_slice(),
-            passphrase.as_str(),
-        )?;
+        // Generate the MAC
+        let expected_mac = Self::generate_mac(&birthday_bytes, entropy.as_ref(), version, &salt, mac_key.as_ref())?;
 
-        if decrypted_mac != mac {
+        // Verify the MAC in constant time to avoid leaking data
+        if mac.ct_eq(&expected_mac).unwrap_u8() == 0 {
             return Err(KeyManagerError::DecryptionFailed);
         }
 
-        let salt_vec: ArrayVec<_, CIPHER_SEED_SALT_BYTES> = salt.into_iter().collect();
-        let salt_bytes = salt_vec.into_inner().map_err(|_| KeyManagerError::InvalidData)?;
-
         Ok(Self {
-            version: received_version,
-            birthday: decrypted_birthday,
-            entropy: decrypted_entropy,
-            salt: salt_bytes,
+            version,
+            birthday,
+            entropy: (*entropy).clone(),
```

Comment:
See other comment. Same question.

Reply:
Good point about early exists. I'd much rather ensure that secret data is wrapped as early as possible so we don't have to worry about it.